### PR TITLE
[Qwen3-TTS] Batch incremental Codec decoding on arena-backed state

### PR DESCRIPTION
**Pull request by BruceLoveDecimal**
## Motivation

T-PR8 of #1754. Stacked on #1756 and #1757.

## Modifications

- **Decoder core (`incremental_codec.py`, from #1756)**
  - Per-row absolute positions in the incremental Transformer: the mask was built from batch-shared 1-D vectors and applied as `view(1, 1, Q, K)`, so B=1 was baked into the mask, not merely guarded. Now `[B, Q]` / `[B, K]`.
  - Retained K/V length comes from the buffer instead of the tracked context length, and keys at negative nominal positions are masked — this is what lets a cold and a warm stream share one cohort with a single execution shape.
  - `state_spec()` / `init_state()` / `state_bytes_per_stream()`, since the lazily grown buffers only materialize after activations flow and an arena cannot preallocate from them.

- **Arena (`codec_state_arena.py`, new)**
  - `Qwen3TTSCodecStateArena`: preallocated `[S, ...]` storage sized from the state spec, with `gather()` / `scatter()` for a cohort.
  - `acquire()` zeroes before handing a slot out, so a reused slot is provably a cold start.
  - `retire()` withdraws a slot whose decode could not be proven complete, mirroring the existing `_CONTEXT_FATAL_RETAINED` precedent for pinned staging.

- **Scheduling (`streaming_vocoder.py`)**
  - The incremental path runs on the async workers again — [#1757](https://github.com/sgl-project/sglang-omni/issues/1757) forced `_async_decode = False` whenever the flag was on — and follow-up decodes cohort by `fresh_frames`; bootstrap decodes stay B=1 because `ref_frames + initial_chunk` is ragged per request.
  - Slots are claimed under `_state_lock` at planning time and released exactly once per cohort after the decode resolves, so an abort landing mid-decode defers the release instead of recycling memory the GPU may still be writing.
  - Failures degrade rather than abort: bad codec rows fail only their own stream (`resolve_partial()` keeps the survivors' deltas, which `resolve()` cannot since a cohort's state has already advanced), and any other decode failure or an exhausted arena falls that request back to the left-context decoder and re-queues it.

## Tests
`tests/unit_test/qwen3_tts/`: **250 passed** on GPU.

### Hardware / stack

- 1× RTX PRO 6000 Blackwell Server Edition (96 GB)
- torch 2.13.0+cu130, sglang 0.5.18, transformers 5.12.1, flashinfer 0.6.17
- Model: `Qwen3-TTS-12Hz-1.7B-Base` in BF16

### Workload

- 100 utterances from the EN split of `zhaochenyang20/seed-tts-eval-arrow` (1088 total), voice cloning from each prompt, `--seed 42`
- WER via `Qwen/Qwen3-ASR-1.7B` (the tooling default)
- Speaker similarity via `popsoda2002/seedtts-wavlm-sim` over the `s3prl/converted_ckpts` WavLM-Large base
- UTMOS via `balacoon/utmos`
- 100/100 completed, with 0 failed requests and 0 left-context fallbacks in every cell

### Test conditions

Quality and speed are measured in two separate rounds, because the settings that make an A/B attributable are not the settings that give representative serving numbers.

**Quality round — deterministic Talker.** At concurrency > 1 a pinned `--seed` is not enough: batch composition follows request timing, so the two vocoder paths make the Talker sample different token sequences. Deterministic decoding pins the token stream, so both arms decode identical codec tokens and any delta belongs to the codec path. The pipeline-level flag would also force the vocoder to B=1 and drop its CUDA graphs — removing the batched path under test — so the vocoder is overridden back at the stage level:

```bash
--enable_deterministic_inference true \
--vocoder.factory.enable_deterministic_inference false \
--vocoder.factory.initial_cuda_graph true \
--vocoder.factory.followup_cuda_graph true
```

With this on, 0 of 100 samples differ in audio length between the arms at c=4 or c=8, confirming the token streams match.

**Performance round — stock configuration**, no flags, so RTF and latency are not distorted by deterministic-mode overhead. Only speed is read from this round; only quality is read from the other.

Scoring is not a confounder: transcribing the same 1088 WAVs twice yields an identical corpus WER (0.9797 %) and 0 differing per-sample WER.

### Quality — deterministic Talker, identical codec tokens

| c | arm | WER corpus % | WER mean % | SIM | UTMOS | UTMOS p5 |
|---|---|---|---|---|---|---|
| 1 | base | 0.728 | 0.710 | 69.73 | 4.148 | 3.542 |
| 1 | this PR | 0.728 | 0.710 | 69.73 | 4.148 | 3.542 |
| 4 | base | 0.728 | 0.710 | 69.73 | 4.148 | 3.542 |
| 4 | this PR | 0.728 | 0.710 | 69.71 | 4.149 | 3.542 |
| 8 | base | 0.728 | 0.710 | 69.73 | 4.149 | 3.542 |
| 8 | this PR | 0.728 | 0.710 | 69.73 | 4.148 | 3.542 |

WER is identical to the digit at every concurrency. SIM moves by at most 0.02 and UTMOS by at most 0.001, which is the batched-matmul reduction-order residual and not a path difference. The baseline's own three rows are also identical to each other, which is the control that shows deterministic decoding took effect.

### Performance — stock configuration

| c | arm | RTF mean | latency mean (s) | p95 (s) | p99 (s) | req/s | audio s/s |
|---|---|---|---|---|---|---|---|
| 1 | base | 0.143 | 0.549 | 0.792 | 0.924 | 1.82 | 7.14 |
| 1 | this PR | 0.141 | 0.545 | 0.782 | 0.907 | 1.83 | 7.20 |
| 4 | base | 0.189 | 0.711 | 1.056 | 1.197 | 5.60 | 21.61 |
| 4 | this PR | 0.190 | 0.716 | 1.036 | 1.172 | 5.55 | 21.38 |
| 8 | base | 0.215 | 0.810 | 1.255 | 1.471 | 9.58 | 36.90 |
| 8 | this PR | 0.222 | 0.837 | 1.307 | 1.527 | 9.23 | 35.52 |

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

### CODE REVIEW

**Comment by JiaxinD** [comment]
Synced the two fixes from the now-green #1757 parent and current `main` onto this draft. The arena-specific `codec_slot` path was preserved while resolving the parent and add/add conflicts. Current head: `e6e95532`. Local validation: `30 passed` in `test_incremental_codec.py`; pre-commit and compileall pass. This remains draft pending the formal benchmark and exact-head GPU/serving validation.

**Comment by JiaxinD** [comment]
@BruceLoveDecimal I want to acknowledge a process mistake on my side. While reviewing this draft, I used maintainer edit access to push directly to your branch without coordinating with you first. I should have prepared the stack update on my fork and offered it separately. Sorry about that.

The branch moved from `ac83822c` to `e6e95532`. I brought in the two parent fixes `e56b3b40` and `8b5c0716` from #1757, then added merge commit `e6e95532` to sync current main while preserving the arena-specific path. I did not force-push, so your original head remains in the history. I believe the stack update is useful, but whether to keep it is your decision.

I will not push to this branch again. If you prefer the original branch state, please say so; I can prepare a fork-based patch or restoration instructions, but I will not change the branch without your explicit request.

**Comment by BruceLoveDecimal** [comment]
> @BruceLoveDecimal I want to acknowledge a process mistake on my side. While reviewing this draft, I used maintainer edit access to push directly to your branch without coordinating with you first. I should have prepared the stack update on my fork and offered it separately. Sorry about that.
> 
> The branch moved from `ac83822c` to `e6e95532`. I brought in the two parent fixes `e56b3b40` and `8b5c0716` from #1757, then added merge commit `e6e95532` to sync current main while preserving the arena-specific path. I did not force-push, so your original head remains in the history. I believe the stack update is useful, but whether to keep it is your decision.
> 
> I will not push to this branch again. If you prefer the original branch state, please say so; I can prepare a fork-based patch or restoration instructions, but I will not change the branch without your explicit request.

Hi @JiaxinD No worries at all :) !— managing my dev branch with an agent makes this kind of thing easy nowadays

Since [#1756](https://github.com/sgl-project/sglang-omni/issues/1756) and [#1757](https://github.com/sgl-project/sglang-omni/issues/1757) are both merged into main now, I've rebased the branch directly onto main.  And I'll rerun the e2e quality and performance numbers and then update the PR body. If those come back clean I'll  mark this PR ready for review.

File: sglang_omni/models/qwen3_tts/streaming_vocoder.py [hunk 1]
```diff
@@ -490,15 +569,29 @@ def __init__(
...
             False
-            if self._enable_stateful_codec_decoder
+            if (self._enable_stateful_codec_decoder and self._deterministic_inference)
             else (
                 self._device.type == "cuda"
                 if async_decode is None
                 else bool(async_decode)
             )
         )
+        self._codec_arena = self._build_codec_arena(int(codec_state_slots))
```
Comment (JiaxinD) [inline sglang_omni/models/qwen3_tts/streaming_vocoder.py]: When stateful decoding runs synchronously (`async_decode=False`, including deterministic mode), this arena is never consumed: `decode_delta()` keeps the per-request state path and no async worker calls `_use_incremental_path()`. The unconditional 64-slot allocation therefore reserves avoidable device memory. I prepared a fork-only patch that allocates the arena only when `_async_decode` is true and adds a regression contract: https://github.com/JiaxinD/sglang-omni/commit/e894084868e12a4b4bdb7feacf981ebcda421e36

File: sglang_omni/models/qwen3_tts/incremental_codec.py [hunk 2]
```diff
@@ -10,6 +10,39 @@
...
+            channels * length for _, channels, length in self.transconv_overlaps
+        )
+        elements += (
+            2
+            * self.num_layers
+            * self.num_key_value_heads
+            * self.retained_context
+            * self.head_dim
+        )
+        return int(elements * itemsize)
```
Comment (JiaxinD) [inline sglang_omni/models/qwen3_tts/incremental_codec.py]: `bytes_per_stream()` omits the per-slot `frame_positions` tensor allocated by `init_state()` (`torch.long`, 8 bytes per stream), so `bytes_per_slot` and `total_bytes` under-report the actual arena allocation. The same fork-only patch counts it and verifies the reported value against every allocated state tensor; the focused incremental-codec suite passes 31/31: https://github.com/JiaxinD/sglang-omni/commit/e894084868e12a4b4bdb7feacf981ebcda421e36

**Comment by BruceLoveDecimal** [comment]
Thanks @JiaxinD — both findings are real, and I've applied your patch.
I cherry-picked `e894084` onto the branch as `80ebb390`, with you as co-author.

**Review by luojiaxuan — COMMENTED** [review]
Reviewed the full stack (arena + incremental_codec batching + streaming_vocoder cohort lifecycle) at high effort. **The core is correct and well-built** — the per-row causal mask ([B,Q,K] with keys/queries broadcast per row), the negative-position masking that lets a right-aligned cold-stream buffer share one execution shape with warm streams, and the buffer-derived prior length are all sound; I traced the partially-warm case (frame_position < retained_context) and the right-alignment + negative mask handles it correctly. The slot lifecycle is the part reviewers will worry about most, and it holds up: in-flight marking under _state_lock, deferred release on mid-decode abort, release-only-after-resolve (which synchronizes), retire-on-unproven-completion, and fallback-to-left-context all route through the pre-existing clear_stream_state -> release_stream_resources seam. No leak or double-advance found across the bad-row, abort, and fallback paths.

Three non-blocking findings:

1. **Dead `frame_positions` device round-trip** (arena.py): `gather()` reads `storage.frame_positions` into the cohort state, but `_decode_incremental_cohorts` overwrites it with a host-built tensor on the very next line; `_zero_slot` and `scatter` keep maintaining the device copy that nothing reads. The host mirror `codec_frame_position` is the only value used. Dropping `frame_positions` from the arena's storage/gather/scatter/zero would remove the redundancy and the misleading 'device positions matter' invariant.

2. **Full `torch.cat(state.code_chunks)` per plan** (streaming_vocoder.py `_build_incremental_plan`): the whole retained code buffer is concatenated on every decode before a small tail slice; slicing the needed window directly avoids one concat+copy per stream per step on the hot path. Bounded by pruning, so minor.

3. **`advance()` bumps the scalar `frame_position` on cohort states** (incremental_codec.py) where it starts at 0 and is never read — harmless now, but a latent trap if a future reader consults the scalar on a cohort. Consider leaving it untouched when `frame_positions` is set.

None block merge. This is the right throughput lever for the 10-20 RPS gap (I verified separately that scheduling-side pacing is not — data on #1903), so I'd like to see this land; #1855 stacks on it. Happy to re-review after the cleanups.

**Comment by luojiaxuan** [comment]
One measured data point for context (not a blocker, and not a criticism of this PR — it does what it claims for the stateful path). Same tree (main + #1852 + this PR), same host/admission, toggling only `vocoder.factory.enable_stateful_codec_decoder` at r10 open-loop:

- default `chunked_decode`: 94.2% success, 2590 s audio generated in the window (~44x realtime aggregate)
- stateful incremental Codec, arena-batched (this PR): 25.8% success, 719 s generated (~12x realtime), 0 arena fallbacks

So the arena batching works as designed (arena engaged, never exhausted), but the stateful path's aggregate Codec throughput is ~3.6x below the default `chunked_decode` at this concurrency, which shows up as most requests missing their first-audio window. Flagging so the stateful path isn't positioned as a high-concurrency throughput win — its advantage is elsewhere (per-request incremental state / correctness properties). If there's a config regime where the stateful path is expected to beat `chunked_decode` at load, I'm happy to re-measure it. My review findings above still stand.

**Comment by BruceLoveDecimal** [comment]
> Reviewed the full stack (arena + incremental_codec batching + streaming_vocoder cohort lifecycle) at high effort. **The core is correct and well-built** — the per-row causal mask ([B,Q,K] with keys/queries broadcast per row), the negative-position masking that lets a right-aligned cold-stream buffer share one execution shape with warm streams, and the buffer-derived prior length are all sound; I traced the partially-warm case (frame_position < retained_context) and the right-alignment + negative mask handles it correctly. The slot lifecycle is the part reviewers will worry about most, and it holds up: in-flight marking under _state_lock, deferred release on mid-decode abort, release-only-after-resolve (which synchronizes), retire-on-unproven-completion, and fallback-to-left-context all route through the pre-existing clear_stream_state -> release_stream_resources seam. No leak or double-advance found across the bad-row, abort, and fallback paths.
> 
> Three non-blocking findings:
> 
> 1. **Dead `frame_positions` device round-trip** (arena.py): `gather()` reads `storage.frame_positions` into the cohort state, but `_decode_incremental_cohorts` overwrites it with a host-built tensor on the very next line; `_zero_slot` and `scatter` keep maintaining the device copy that nothing reads. The host mirror `codec_frame_position` is the only value used. Dropping `frame_positions` from the arena's storage/gather/scatter/zero would remove the redundancy and the misleading 'device positions matter' invariant.
> 2. **Full `torch.cat(state.code_chunks)` per plan** (streaming_vocoder.py `_build_incremental_plan`): the whole retained code buffer is concatenated on every decode before a small tail slice; slicing the needed window directly avoids one concat+copy per stream per step on the hot path. Bounded by pruning, so minor.
> 3. **`advance()` bumps the scalar `frame_position` on cohort states** (incremental_codec.py) where it starts at 0 and is never read — harmless now, but a latent trap if a future reader consults the scalar on a cohort. Consider leaving it untouched when `frame_positions` is set.
> 
> None block merge. This is the right throughput lever for the 10-20 RPS gap (I verified separately that scheduling-side pacing is not — data on #1903), so I'd like to see this land; #1855 stacks on it. Happy to re-review after the cleanups.

Thanks for the thorough pass! @luojiaxuan 
**(1) and (3) — agreed, fixed.**
**(2)** I think It is a clarity problem more than a throughput one, which matches your "bounded by pruning, so minor". 
The reason I would rather not fold it in here: the concat is quietly buying us a private copy. `validate_chunk` does `codes.detach().to(dtype=torch.long)`, which returns the same tensor when the input is already `long`, so `state.code_chunks` may alias the engine's buffer.

**Comment by BruceLoveDecimal** [comment]
> One measured data point for context (not a blocker, and not a criticism of this PR — it does what it claims for the stateful path). Same tree (main + #1852 + this PR), same host/admission, toggling only `vocoder.factory.enable_stateful_codec_decoder` at r10 open-loop:
> 
> * default `chunked_decode`: 94.2% success, 2590 s audio generated in the window (~44x realtime aggregate)
> * stateful incremental Codec, arena-batched (this PR): 25.8% success, 719 s generated (~12x realtime), 0 arena fallbacks
> 
> So the arena batching works as designed (arena engaged, never exhausted), but the stateful path's aggregate Codec throughput is ~3.6x below the default `chunked_decode` at this concurrency, which shows up as most requests missing their first-audio window. Flagging so the stateful path isn't positioned as a high-concurrency throughput win — its advantage is elsewhere (per-request incremental state / correctness properties). If there's a config regime where the stateful path is expected to beat `chunked_decode` at load, I'm happy to re-measure it. My review findings above still stand.

Agreed — this PR is not a performance win. This PR is groundwork for T-PR9.

The cause is in this PR. With `enable_stateful_codec_decoder=True` both graph sets are
disabled — `streaming_vocoder.py:534` and `:553` each carry
`and not self._enable_stateful_codec_decoder`. My smoke confirms it: base
`graphs=2 arena=0`, stateful `graphs=0 arena=1`.

So the comparison is base at 24 frames per decode on a captured graph, versus this PR
at 8 frames fully eager. Low concurrency, the saved compute wins; at load the launch
overhead accumulates linearly and outweighs the saved FLOPs. That is T-PR9 — graph
coverage for incremental shapes, as the comment at `:553` says.

But it's hypothesis, not a measurement — I haven't isolated per-decode launch overhead. Happy
to take the re-measure offer once incremental shapes are captured.

**Comment by luojiaxuan** [comment]
Qualified this together with #1855 on the r20 harness and rebased both on current main; findings and the branch are in https://github.com/sgl-project/sglang-omni/pull/1855#issuecomment-5559304161.

**Comment by luojiaxuan** [comment]
#1997 carries this PR's commits merged with main plus the follow-up fixes and the CUDA graph work from #1855, validated end to end (rps 20 underrun 20.9% to ~1.2%, first playable p50 89 to 56 ms, 100% completion). Suggest reviewing there; happy to split it back out if you prefer this one to land first.

**Comment by BruceLoveDecimal** [comment]
> #1997 carries this PR's commits merged with main plus the follow-up fixes and the CUDA graph work from #1855, validated end to end (rps 20 underrun 20.9% to ~1.2%, first playable p50 89 to 56 ms, 100% completion). Suggest reviewing there; happy to split it back out if you prefer this one to land first.
> 

Either way works for me, whichever is more convenient for you. I’ll keep track of #1997’s progress. If any changes are needed on my side, or if there are any questions, feel free to ping me in the WeChat group. Anyway， thanks for keeping track of my PR!

**Comment by BruceLoveDecimal** [comment]
Since #1997 has been merged already, it can be closed now.
